File: trunk/server/src/logic/prpc/COM_BattleReport.go
package prpc
import(
  "bytes"
  "encoding/json"
  "suzuki/prpc"
)
type COM_BattleReport struct{
  BattleID int32  //0
  UnitList []COM_BattleUnit  //1
  ActionList []COM_BattleAction  //2
}
func (this *COM_BattleReport)Serialize(buffer *bytes.Buffer) error {
  //field mask
  mask := prpc.NewMask1(1)
  mask.WriteBit(this.BattleID!=0)
  mask.WriteBit(len(this.UnitList) != 0)
  mask.WriteBit(len(this.ActionList) != 0)
  {
    err := prpc.Write(buffer,mask.Bytes())
    if err != nil {
      return err
    }
  }
  // serialize BattleID
  {
    if(this.BattleID!=0){
      err := prpc.Write(buffer,this.BattleID)
      if err != nil{
        return err
      }
    }
  }
  // serialize UnitList
  if len(this.UnitList) != 0{
    {
      err := prpc.Write(buffer,uint(len(this.UnitList)))
      if err != nil {
        return err
      }
    }
    for _, value := range this.UnitList {
      err := value.Serialize(buffer)
      if err != nil {
        return err
      }
    }
  }
  // serialize ActionList
  if len(this.ActionList) != 0{
    {
      err := prpc.Write(buffer,uint(len(this.ActionList)))
      if err != nil {
        return err
      }
    }
    for _, value := range this.ActionList {
      err := value.Serialize(buffer)
      if err != nil {
        return err
      }
    }
  }
  return nil
}
func (this *COM_BattleReport)Deserialize(buffer *bytes.Buffer) error{
  //field mask
  mask, err:= prpc.NewMask0(buffer,1);
  if err != nil{
    return err
  }
  // deserialize BattleID
  if mask.ReadBit() {
    err := prpc.Read(buffer,&this.BattleID)
    if err != nil{
      return err
    }
  }
  // deserialize UnitList
  if mask.ReadBit() {
    var size uint
    err := prpc.Read(buffer,&size)
    if err != nil{
      return err
    }
    this.UnitList = make([]COM_BattleUnit,size)
    for i,_ := range this.UnitList{
      err := this.UnitList[i].Deserialize(buffer)
      if err != nil{
        return err
      }
    }
  }
  // deserialize ActionList
  if mask.ReadBit() {
    var size uint
    err := prpc.Read(buffer,&size)
    if err != nil{
      return err
    }
    this.ActionList = make([]COM_BattleAction,size)
    for i,_ := range this.ActionList{
      err := this.ActionList[i].Deserialize(buffer)
      if err != nil{
        return err
      }
    }
  }
  return nil
}
func (this *COM_BattleReport)String() string{
  b, e := json.Marshal(this)
  if e != nil{
    return e.Error()
  }else{
    return string(b)
  }
}

File: trunk/Client/Assets/Scripts/Common/Rpc/COM_BattleReport.cs
public class COM_BattleReport{
  public virtual void Serialize(IWriter w){
    Mask mask = new Mask(1);
    mask.WriteBit(UnitList!=null&&UnitList.Length!=0);
    mask.WriteBit(ActionList!=null&&ActionList.Length!=0);
    w.Write(mask.Bytes);
    //S UnitList
    if(UnitList!=null&&UnitList.Length!=0){
      w.WriteSize(UnitList.Length);
      for(int i=0; i<UnitList.Length; ++i){
        UnitList[i].Serialize(w);
      }
    }
    //S ActionList
    if(ActionList!=null&&ActionList.Length!=0){
      w.WriteSize(ActionList.Length);
      for(int i=0; i<ActionList.Length; ++i){
        ActionList[i].Serialize(w);
      }
    }
  }
  public virtual bool Deserialize(IReader r){
    Mask mask = new Mask(1);
    if(!r.Read(ref mask.Bytes)){
      return false;
    }
    //D UnitList
    if(mask.ReadBit()){
      int size = 0;
      if(!r.ReadSize(ref size) || size > 255){
        return false;
      }
      UnitList = new COM_BattleUnit[size];
      for(int i=0; i<size; ++i){
        UnitList[i] = new COM_BattleUnit();
        if(!UnitList[i].Deserialize(r)){
          return false;
        }
      }
    }
    //D ActionList
    if(mask.ReadBit()){
      int size = 0;
      if(!r.ReadSize(ref size) || size > 255){
        return false;
      }
      ActionList = new COM_BattleAction[size];
      for(int i=0; i<size; ++i){
        ActionList[i] = new COM_BattleAction();
        if(!ActionList[i].Deserialize(r)){
          return false;
        }
      }
    }
    return true;
  }
  public COM_BattleUnit[] UnitList;
  public COM_BattleAction[] ActionList;
}
